scripts/tsrc/TSRC.ts

```diff
@@ -0,0 +1,408 @@
+	 */
+	constructor(
+		private readonly projectDirectory: string,
+		virtualModuleName: string,
+		ignore: string[] = []
+	) {
+		// Interpret relative paths as being relative to projectDirectory.
+		// Calling `resolve` has no effect if `ignorePath` is absolute, so it's safe to call it unconditionally.
+		this.outDirectory = resolve(projectDirectory, virtualModuleName);
+		this.ignore = ignore.map((path) => resolve(projectDirectory, path));
+		this.virtualModuleName = virtualModuleName;
+
+		// eslint-disable-next-line @typescript-eslint/unbound-method
+		const tsconfigPath = ts.findConfigFile(this.projectDirectory, ts.sys.fileExists);
+
+		if (!tsconfigPath) {
+			throw new Error("Could not find a valid 'tsconfig.json'.");
+		}
+		this.tsconfigPath = tsconfigPath;
+
+		// eslint-disable-next-line @typescript-eslint/unbound-method
+		const result = ts.readConfigFile(tsconfigPath, ts.sys.readFile);
+		if (result.error) {
+			throw new Error(result.error.messageText.toString());
+		}
+
+		this.tsconfig = ts.parseJsonConfigFileContent(result.config, ts.sys, projectDirectory);
+
+		this.program = ts.createProgram(this.tsconfig.fileNames, this.tsconfig.options);
+	}
+
+	private ajv = new Ajv({
+		allowUnionTypes: true,
+		code: {
+			// Add line breaks to generated code
+			lines: true,
+			source: true,
+			esm: true,
+			optimize: true,
+		},
+	});
+
+	public readonly tsconfigPath: string;
+
+	private tsconfig;
+
+	public readonly reporter = new GeneratedFilesReporter();
+
+	private program: ts.Program;
+
+	processSourceFiles() {
+		for (const sourceFile of this.program.getSourceFiles()) this.processSourceFile(sourceFile);
+	}
+
+	/**
+	 * Traverses the AST of `sourceFile` for type declarations and imports, and extracts the types that are eligible for
+	 * runtime type checks.
+	 */
+	private processSourceFile(sourceFile: ts.SourceFile) {
+		// Skip files in node_modules
+		if (sourceFile.fileName.includes("/node_modules/")) return;
+		// Skip files in the output directory
+		if (sourceFile.fileName.startsWith(this.outDirectory)) return;
+
+		for (const ignorePath of this.ignore) {
+			if (sourceFile.fileName.startsWith(resolve(this.projectDirectory, ignorePath))) return;
+		}
+
+		const moduleName = this.moduleName(sourceFile);
+		let types = this.typeDecl.get(moduleName);
+
+		sourceFile.forEachChild((node) => {
+			// Detect which declarations have been used
+			if (ts.isImportDeclaration(node)) {
+				// We only care about imports from the special virtual module
+				// Get the module specifier without quotes
+				const module = node.moduleSpecifier.getFullText(sourceFile).trim().slice(1, -1);
+				// Append a slash to the virtual module name to avoid matching modules with a coincidental
+				// common prefix
+				if (!module.startsWith(`${this.virtualModuleName}/`)) return;
+				// Omit imports declarations that don't import anything
+				if (!node.importClause) return;
```

Comment:
If this call is faster than `node.moduleSpecifier.getFullText()` we should consider swapping the order of these returns.

Reply:
Fixed